rfcs/20230622-quantized-reduction.md

```diff
@@ -0,0 +1,521 @@
+        %add = "stablehlo.add"(%elem_rescaled, %acc)
+            : (tensor<!quant.uniform<i32:f32, input_scale:0>>,
+               tensor<!quant.uniform<i32:f32, input_scale:0>>)
+            -> tensor<!quant.uniform<i32:f32, input_scale:0>>
+        "stablehlo.return"(%0)
+            : (tensor<!quant.uniform<i32:f32, input_scale:0>>) -> ()
+    }) {
+        dimensions = dense<1> : tensor<1xi64>
+    } : (tensor<... x !quant.uniform<ui8:f32, input_scale:input_zp>>,
+         tensor<... x !quant.uniform<i32:f32, input_scale:0>>)
+    -> tensor<... x !quant.uniform<i32:f32, input_scale:0>>
+```
+
+In this option, the `init_value` type and the `result` type can be different
+from the input type. The first argument of the compute block is fixed for the
+traversed element and the second argument is fixed for the intermediate
+(accumulation) result.
+
+#### Tradeoffs
+
+* (+) Make the accumulation type explicit in the IR.
+* (-) This representation imposes a limitation on the evaluation order.
+  Since we can’t express the computation between two intermediate (accumulation)
+  results, we can not arbitrarily insert `init_value` and start the
+  computation at an arbitrary location. The following shows the restricted
+  evaluation order with the method.
+
+```python
++----------+       +--------+ +--------+ +--------+ +--------+
+|init_value|       |input[0]| |input[1]| |input[2]| |input[3]|
++----------+       +--------+ +--------+ +--------+ +--------+
+           \        /         /          /          /
+           +-------+         /          /          /
+           |compute|        /          /          /
+           +-------+       /          /          /
+                \         /          /          /
+                 +-------+          /          /
+                 |compute|         /          /
+                 +-------+        /          /
+                       \         /          /
+                        +-------+          /
+                        |compute|         /
+                        +-------+        /
+                              \         /
+                               +-------+
+                               |compute|
+                               +-------+
+```
+
+## Open Question
+
+### Should we restrict the proposal #1 to quantized types only?
+
+The above proposal #1 of introducing the additional functions is theoretically
+not limited to quantized `reduce` op, but also can be applied to `reduce` op with
+non-quantized types. For example,
+
+```mlir
+%result = "stablehlo.reduce"(%input, %init_value) ({
+  ^input_conversion(%arg0: tensor<bf16>):
+    %0 = "stablehlo.convert"(%arg0): (tensor<bf16>) -> (tensor<f32>)
+    "stablehlo.return"(%0) : (tensor<f32>) -> (tensor<f32>)
+  }, {
+  ^bb0(%arg0: tensor<f32>, %arg1: tensor<f32>):
+    %0 = "stablehlo.add"(%arg0, %arg1) : (tensor<f32>, tensor<f32>) ->
+    tensor<f32>
+    "stablehlo.return"(%0) : (tensor<f32>) -> ()
+  },
```

Comment:
Missing a `{` at the end of line to start a region

Reply:
done.